pkg/resource/deploy/deployment_executor.go

```diff
@@ -43,25 +42,28 @@ type deploymentExecutor struct {
 // checkTargets validates that all the targets passed in refer to existing resources.  Diagnostics
 // are generated for any target that cannot be found.  The target must either have existed in the stack
 // prior to running the operation, or it must be the urn for a resource that was created.
-func (ex *deploymentExecutor) checkTargets(targets UrnTargets, op display.StepOp) result.Result {
+func (ex *deploymentExecutor) checkTargets(targets UrnTargets) result.Result {
 	if !targets.IsConstrained() {
 		return nil
 	}
 
 	olds := ex.deployment.olds
 	var news map[resource.URN]bool
+	var deletes map[resource.URN]bool
 	if ex.stepGen != nil {
 		news = ex.stepGen.urns
+		deletes = ex.stepGen.deletes
 	}
 
 	hasUnknownTarget := false
 	for _, target := range targets.Literals() {
 		hasOld := olds != nil && olds[target] != nil
 		hasNew := news != nil && news[target]
-		if !hasOld && !hasNew {
+		hasDelete := deletes != nil && deletes[target]
+		if !hasOld && !hasNew && !hasDelete {
```

Comment:
deleted resources don't show up in olds. not sure if this is intentional.

Reply:
This is odd because we call checkTargets before iterating the source so at that point I'd think news and deletes are empty because we haven't actually run the step generator yet?

Reply:
We're now checking deployments after iterating the source. This is to make tests pass since the pre-check is now performed much earlier in a component above the engine similarly to --exclude-protected.

There is some desire to not fail a targeted destroy if a target does not exist.
- https://github.com/pulumi/pulumi/issues/8562

As a sort of, ensure this no longer exists.

Reply:
Right yeh ok so this gets called after the update, but it used to be called before the update for destroy/refresh but that's now handled by a copy of this code.
Still really confused why deleted resources aren't in olds? I can't find where we mutate old to remove stuff, the changes to add the `deletes` map here feel like they shouldn't be needed.

Reply:
This seems to explain it

https://github.com/pulumi/pulumi/blob/df02d6ded7c44c8b6251559eaece86c4811cbb9e/pkg/resource/deploy/deployment.go#L401-L404

https://github.com/pulumi/pulumi/blob/df02d6ded7c44c8b6251559eaece86c4811cbb9e/sdk/go/common/resource/resource_state.go#L33

Reply:
Oh those aren't deletes, those are pending deletes... I'm not sure how --targets should handle that 🤔

Reply:
I've figured it out. It's because we overwrite ex.deployment.olds.


this is interesting, we replace olds on targeted destroys because of the side effects when running `rebuildBaseState()`  this was added initially for refreshes
https://github.com/pulumi/pulumi/blob/ef69ffba3ab552265577e9878578002e5e9ca9d8/pkg/resource/deploy/deployment_executor.go#L609

some discussion on the earliest addition of this: https://github.com/pulumi/pulumi/pull/1814/files#r212073292

It ended up being run on targeted destroys
https://github.com/pulumi/pulumi/blob/41996a3cbd145fbd1ce865b41a753136223bec1c/pkg/resource/deploy/plan_executor.go#L233-L242